examples/grad3d/grad3d.f90

```diff
@@ -0,0 +1,625 @@
+   ! Calculate gradient in Y (data in Y-pencil)
+   !=====================================================================
+   subroutine dery(df,ff,delta,nx,ny,nz)
+
+      implicit none
+
+      ! Arguments
+      integer, intent(in) :: nx, ny, nz
+      real(mytype), intent(in) :: delta
+      real(mytype), intent(out), dimension(nx,ny,nz) :: df
+      real(mytype), intent(in), dimension(nx,ny,nz) :: ff
+
+      ! Local variables
+      integer :: i, j, k
+      real(mytype) :: coeff = 0.5_mytype
+
+      coeff = coeff/delta
+
+      !$acc kernels default(present)
+      do concurrent (k=1:nz)
+        do concurrent (i=1:nx)
+           df(i,1,k) = coeff*(ff(i,2,k)-ff(i,ny,k))
+        enddo
+        do concurrent (j=2:ny-1, i=1:nx)
+           df(i,j,k) = coeff*(ff(i,j+1,k)-ff(i,j-1,k))
+        enddo
+        do concurrent (i=1:nx)
+           df(i,ny,k) = coeff*(ff(i,1,k)-ff(i,ny-1,k)) 
+        enddo
+      enddo
+      !$acc end kernels
+
+   end subroutine dery 
+   !=====================================================================
+   ! Calculate gradient in Z (data in Z-pencil)
+   !=====================================================================
+   subroutine derz(df,ff,delta,nx,ny,nz)
+
+      implicit none
+
+      ! Arguments
+      integer, intent(in) :: nx, ny, nz
+      real(mytype), intent(in) :: delta
+      real(mytype), intent(out), dimension(nx,ny,nz) :: df
+      real(mytype), intent(in), dimension(nx,ny,nz) :: ff
+
+      ! Local variables
+      integer :: i, j, k
+      real(mytype) :: coeff = 0.5_mytype
+      
+      coeff = coeff/delta
+
+      !$acc kernels default(present)
+      do concurrent (j=1:ny, i=1:nx)
+         df(i,j,1) = coeff*(ff(i,j,2)-ff(i,j,nz  ))
+      enddo
+      do concurrent (k=2:nz-1, j=1:ny, i=1:nx)
+         df(i,j,k) = coeff*(ff(i,j,k+1)-ff(i,j,k-1))
+      enddo
+      do concurrent (j=1:ny, i=1:nx)
+         df(i,j,nz) = coeff*(ff(i,j,1)-ff(i,j,nz-1))
+      enddo
+      !$acc end kernels
+
+   end subroutine derz 
+   !=====================================================================
+   ! Calculate divergence using halo-cell exchange (data in X-pencil)
+   !=====================================================================
+   subroutine test_derX(df)
+
+      implicit none
+      ! Arguments
+      real(mytype), intent(out), dimension(xsize(1),xsize(2),xsize(3)) :: df
+
+      integer :: i, j, k
+      real(mytype), parameter :: twopi=2._mytype*acos(-1._mytype)
+      real(mytype) :: x, y, z
+      real(mytype) :: dphi, dphi_num
+      real(mytype) :: error = 0._mytype
+      real(mytype) :: err_all = 0._mytype
+      real(mytype) :: dphi2 = 0._mytype
+      real(mytype) :: sum_dphi2 = 0._mytype
+      integer :: xe1, xe2, xe3
+      integer :: xs1, xs2, xs3
+
+      xe1 = xsize(1)
+      xe2 = xsize(2)
+      xe3 = xsize(3)
+      xs1 = xstart(1)
+      xs2 = xstart(2)
+      xs3 = xstart(3)
+
+      ! Initial velocity
```

Comment:
Compute the error against the analytical solution

Reply:
done